esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -1320,6 +1351,138 @@ void EsdlServiceImpl::getSoapError( StringBuffer& out,
     out.clear().append(finger-start,start);
 }
 
+void EsdlServiceImpl::adjustTargetConfig(Owned<IPTree>& tgtcfg) const
+{
+    static const VStringBuffer gwLocalSecretXPath("Gateways/Gateway[@url='%s*']", gwLocalSecretPrefix);
+
+    if (!tgtcfg)
+        return;
+
+    if (tgtcfg->hasProp(gwLocalSecretXPath))
+    {
+        tgtcfg.setown(createPTreeFromIPT(tgtcfg));
+        const char* qt = tgtcfg->queryProp("@querytype");
+        bool published = (qt && (strieq(qt, "roxie") || strieq(qt, "wsecl")));
+        const char* permission = (published ? "allowPublishedGatewayUsage" : "allowUnpublishedGatewayUsage");
+        Owned<IPTreeIterator> gwsToResolve(tgtcfg->getElements(gwLocalSecretXPath));
+        ForEach(*gwsToResolve)
+            resolveGatewayLocalSecret(gwsToResolve->query(), permission);
+    }
+}
+
+void EsdlServiceImpl::resolveGatewayLocalSecret(IPTree& gateway, const char* permission) const
+{
+    const char* name = gateway.queryProp("@name");
+    StringBuffer url(gateway.queryProp("@url"));
+    if (url.isEmpty())
+        throw makeStringExceptionV(-1, "gateway %s: missing url property", name);
```

Comment:
You commented on @name twice. I'll address here. There is a logical requirement for @name. In the legacy transform, @name becomes @ServiceName (see lines 1477 and 1479). There is a need for something that the roxie can use to select the gateway and we've assumed for years it is called @name.

That's not to say that nothing should change. What should the absence of @name mean? We've been assuming its presence for legacy transforms. Absence here, and at 1414, only affects error reporting. Should we default to something like "unknown" for error reporting, because we don't know there isn't another naming convention in use, but fail when we try to act on it?

Reply:
I was only thinking in terms of the local effect of calling makeStringExceptionV with a  possible null pointer. If it seems overkill to have more checks then you should leave as-is.